aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -216,43 +247,43 @@ private void uploadParts() {
       return;
     }
 
-    stagingFiles.stream()
+    stagingFilesWithETags.stream()
         // do not upload the file currently being written
-        .filter(f -> closed || !f.equals(currentStagingFile))
+        .filter(f -> closed || !f.file().equals(currentStagingFile))
         // do not upload any files that have already been processed
-        .filter(Predicates.not(multiPartMap::containsKey))
+        .filter(Predicates.not(f -> multiPartMap.containsKey(f.file())))
         .forEach(f -> {
           UploadPartRequest.Builder requestBuilder = UploadPartRequest.builder()
               .bucket(location.bucket())
               .key(location.key())
               .uploadId(multipartUploadId)
-              .partNumber(stagingFiles.indexOf(f) + 1)
-              .contentLength(f.length());
+              .partNumber(stagingFilesWithETags.indexOf(f) + 1)
+              .contentLength(f.file().length());
 
           S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
 
           UploadPartRequest uploadRequest = requestBuilder.build();
 
           CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
               () -> {
-                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f.file()));
+                checkEtag(f.eTag(), response.eTag());
```

Comment:
S3 has built in checksum verification using request header `Content-MD5`. I think we don't need to check the response at client side.

See https://aws.amazon.com/premiumsupport/knowledge-center/data-integrity-s3/ for more details

Reply:
Yea, I was also initially doing that. However, I could not find a way to add a reliable test. Tests would succeed with wrong md5 checksums added to request. Likely due to s3 mock. With that and the fact that current approach allows to add a better error message, I would propose doing explicit checks here. However, if you have a strong preference on this, I can modify. Let me know.

Reply:
@jackye1995 thoughts?

Reply:
The integration tests in `/aws/src/integration` are added for such a purpose, I think we can leverage that.

Unfortunately it is not run as a part of CI yet because it requires connection to an AWS account, we are working on getting one for Iceberg. If you do not have any AWS infrastructure that could be used to run it, I can run it for you after you publish you tests in this PR.

Reply:
> With that and the fact that current approach allows to add a better error message, I would propose doing explicit checks here.

Consulted with the S3 team related to this, definitely doing it on server side is preferred as they have dedicated hardware for this operation, comparing to a client running checksum in any environment

Reply:
Thanks @jackye1995 . Let me update the diff in that case. Will try to do so this week.

Reply:
@jackye1995 I think this is good now. I verified that wrong md5 is causing protocol exceptions from s3 server side.